File: classwork/closure/closure.py
def makeAdder(n):
    def inner(x):
        return x + n
    return inner

add3 = makeAdder(3)
add5 = makeAdder(5)

print add3(10)
print add5(10)

def make_counter():
    # private "instance" data
    # has to be a list due to weird python scoping rules
    count[0]

    # public methods
    def inc():
        count[0] = count[0] + 1
    def dec():
        count[0] = count[0] - 1
    def reset():
        count[0] = 0
    def get():
        return count[0]

    # return a dictionary so we have access to all the methods
    return {'inc': inc, 'dec': dec, 'reset': reset, 'get': get}

#c1 = make_counter()
#c2 = make_counter()

#c1['inc']()
#c1['inc']()
#c1['inc']()
#print c1['get']()

import random

# @doubler
#def get_name():
#    names = ['bob','tom','sue']
#    return random.choice(names)

#def dble(f):
#    name = f()
#    return name + name

# print dble(get_name)

def doubler(f):
    def inner():
        name = f()
        return name + name
    return inner

@doubler
def get_name():
    names = ['bob','tom','sue']
    return random.choice(names)

print get_name()

"""
2nd eg will have demo return whenever we invoke it

equiv to get_name = double(get_name)

TAKEAWAY:
Bam!: can write fxns that transform fxns

A Python decorator is merely shorthand for calling a wrapper-type 
function like doubler

A Python decorator encapsulates a closure

A decorator allows to transparently wrap functionality
"""

File: classwork/js_closure/closure.js
var base = "this is in f1 in the global namespace";
var x = "x in the global namespace";
var f2 = function() {
  console.log("this is f2");
};

var f1 = {
  x : "something in the f1 namespace" ,
  f2 : f2,
  f : function() {
    console.log('this is in f in the f1 namespace');
    console.log('x is ' + x);
    console.log('f1.x is ' + f1.x);
    console.log('better way: ' + this.x);
  }
};
// f1.f()


// var count = 0;
//
// var makeIncrementer = function() {
//   console.log(count);
//   count += 1;
// };
//
// var inc = function() {
//   return makeIncrementer();
// };

var makeIncrementer = function() {
  var x = 0;
  return function() {
    x = x + 1;
    return x;
  }
}
// In browser:
// var inc = makeIncrementer()
// inc() --> 1

var makeAdder = function(n) {
  var x = n;

  var inner = function() {
    x += x;
    return x;
  }
  return inner(x);
}

var makeAdder2 = function(n) {
  return function(x) {
    return x + n;
  }
}

var add3 = makeAdder2(3);
// add3(10) --> 13

var makeCounter = function() {
  var i = 0;

  var get = function() {
    return i;
  }

  return {
    x: "this is a str in dict",
    set: function(x) {
      i = x;
    },
    get: get,
  }
}
